src/test/PulsarMock.js

```diff
@@ -0,0 +1,251 @@
+export class PulsarClientMock {
+  // A map used to simulate an in-memory store for messages. The keys are topic names, and the values are arrays of message objects.
+  static inMemoryStore = new Map();
+
+  // A counter to generate unique message IDs. It is incremented each time a message is sent.
+  static messageIdCounter = 0;
+
+  // A map where keys are topic names and values are arrays of receive function invocations (listeners). These listeners are called to notify about new messages.
+  static listeners = new Map();
+
+  // The default time in milliseconds to wait before a message is considered not acknowledged (acknowledged messages are removed from visibility to simulate message acknowledgment behavior).
+  static ackTimeout = 30000;
+
+  // A nested map where the first key is the topic name, the second key is the subscriber ID, and the value is a set of message IDs that have been acknowledged by the subscriber.
+  static acknowledgedMessages = new Map();
+
+  static configureAckTimeout(timeout) {
+    this.ackTimeout = timeout;
+  }
+
+  async createProducer(config) {
+    const topic = config.topic;
+    if (!PulsarClientMock.inMemoryStore.has(topic)) {
+      PulsarClientMock.inMemoryStore.set(topic, []);
+    }
+
+    return {
+      send: async (message) => {
+        const messageId = new MessageId(`message-${++PulsarClientMock.messageIdCounter}`);
+        const messageInstance = new Message(topic, undefined, message.data, messageId, Date.now(), Date.now(), 0, '');
+        const messages = PulsarClientMock.inMemoryStore.get(topic) || [];
+        messages.push({ message: messageInstance, subscriberIds: new Set() }); // Track which subscribers have received the message
+
+        // Ensure all subscribers are aware of the new message
+        PulsarClientMock.notifyListeners(topic);
+
+        return { messageId: messageId.toString() };
+      },
+      close: async () => {},
+    };
+  }
+
+  /**
+   * creates a consumer per topic and subscriberId. If more than one subscriber with the same subscriberId is created
+   * for the same topic, they will process messages in a FIFO manner.
+   */
+
+  async subscribe(config) {
+    const topic = config.topic;
+    const subscriberId = config.subscription;
+
+
+    return {
+
+      /**
+       * receives either returns the next visible message, or blocks until a message is available.
+       */
+      receive: () => {
```

Comment:
since we don't have TS to declare return type is promise, marking this method `async` may be a good idea for documenting that it returns a Promise